### ISSUE TITLE
[BUG] Current OPD TopK reward cannot distill teacher but only maximize student's entropy

### ISSUE THREAD

**Issue by YSLIU627** [radixark/miles#2362]
## Title
OPD top-k KL is applied as an action-independent advantage penalty, not as a distillation loss — on-policy gradient contribution is zero

## Existing Docs and Environment
- Branch/tag: `main` @ `77ca8ab`
- https://miles.radixark.com/docs/advanced/on-policy-distillation
## Summary
`--opd-log-prob-top-k > 0` computes a per-position reverse-KL over the *candidate* top-k token set, then subtracts it from the per-token advantage. Because this value is a deterministic function of the prefix (state) and never depends on the token actually sampled at that position, the standard score-function identity makes its expected gradient contribution exactly zero on-policy. It is therefore not a valid per-token reward — information from the top-k set never reaches the optimizer the way the Rethinking-OPD recipe intends.

## Key code references
1. **Penalty is action-independent** — `miles/rollout/on_policy_distillation.py:316-349`
   ```python
   for i in range(response_length):
       selected_ids = _selected_token_ids(strategy, student_ids, teacher_ids)   # candidate set only
       ...
       reverse_kl = sum(w * (s_logp - t_logp) ...)    # no dependence on the sampled token at position i
   The sampled token (sample.tokens[-response_length+i]) is never read anywhere in this function.
2. Applied as a detached constant penalty on advantages — miles/backends/training_utils/loss_hub/opd.py:33-58
reverse_kl = reverse_kl.detach().to(device=adv.device)   # :48
advantages[i] = adv - args.opd_kl_coef * reverse_kl      # :54
Contrast with the sampled-token path at opd.py:88-90 (old_student_log_prob - teacher_log_probs[i]), which depends on the sampled token and is a legal reward (E[ℓ·∇log π] ≠ 0).
3. Pure-distillation config zeroes the base advantage — miles/rollout/on_policy_distillation.py:416-434 returns rewards 0.0; miles/backends/training_utils/loss_hub/math_utils.py:436-443 (get_grpo_returns) only broadcasts scalars; so after miles/backends/training_utils/loss.py:88-107, advantage is just -λ·rKL with no task signal.
Math

```
E_{a~π_θ}[∇ log π_θ(a|s_t)] = 0
⇒ E[-λ·rKL_t(s_t)·∇ log π_θ(a_t)] = -λ·rKL_t·E[∇ log π_θ(a_t)] = 0
```
The term only picks up a signal through rollout/train off-policyness (E_{π_old}[∇log π_new] = -∇KL(π_old‖π_new)), i.e. it aligns the student with its own previous snapshot rather than with the teacher distribution — unlike the paper's official implementation (thunlp/OPD verl/recipe/gkd/megatron_kl_loss.py), where the KL is computed on current training logits (backward = softmax - target) and top-k information truly enters the loss.
## Suggestion
move top-k KL into the training forward as a differentiable loss (gather current logits at teacher top-k indices, like the paper code in https://arxiv.org/pdf/2604.13016).

**Comment by austin362667** [comment]
Thanks for pointing out. But I'm a bit confused.
Can you share details on which section in Rethinking OPD [paper](https://arxiv.org/pdf/2604.13016) supports the issue you raised?

**Comment by yxliu02** [comment]
Thanks for asking. The strongest support is the combination of the paper's stated objective and its official implementation. My original wording should have made that distinction clearer.

### 1. The paper defines top-k OPD as a subset KL, not a scalar state reward

In **Section 2 (Preliminaries), “Top-k OPD”**, Eq. `opd_topk_obj_prelim` defines

```text
E_{y ~ pi_theta} [ sum_t KL(p_bar_t^(S_t) || q_bar_t^(S_t)) ]
```

where both student and teacher distributions are renormalized on `S_t`. Immediately afterward, the paper says this preserves **multi-token supervision** on the student's high-probability region.

Paper: https://arxiv.org/pdf/2604.13016

That objective requires the optimizer to receive derivatives with respect to the student probabilities for the tokens in `S_t`. Collapsing the KL to one detached scalar per state and attaching it only to the sampled action does not provide that gradient.

### 2. The official implementation keeps the K dimension and recomputes live top-k log-probs

The official implementation makes the intended gradient path explicit:

1. `compute_distillation_reward` produces `rm_scores` with shape `[batch, response_len, K]`; it does **not** sum over K:
   https://github.com/thunlp/OPD/blob/ac26e38d6f1572eb027597b48a9f4e01f6915ef8/verl/verl/workers/actor/dp_actor.py#L452-L623
2. During the actor update, when `advantages.dim() == 3`, it recomputes `topk_log_probs` in the **current training forward** and uses them as `log_prob_for_loss`:
   https://github.com/thunlp/OPD/blob/ac26e38d6f1572eb027597b48a9f4e01f6915ef8/verl/verl/workers/actor/dp_actor.py#L825-L846
3. Those live top-k log-probs and the K-dimensional advantages are passed together to the policy-loss function:
   https://github.com/thunlp/OPD/blob/ac26e38d6f1572eb027597b48a9f4e01f6915ef8/verl/verl/workers/actor/dp_actor.py#L880-L900

So even though the official code expresses top-k OPD through a policy-loss interface, every selected token receives a gradient. That is materially different from Miles' current path.

### 3. Miles collapses K before training and differentiates only the sampled action

Miles computes

```python
reverse_kl = sum(w * (s_logp - t_logp) for ...)
```

for each position, storing one scalar in `sample.opd_reverse_kl`:
https://github.com/radixark/miles/blob/ca25d697c0743fde52d27a5ebd88a3883b7fd838/miles/rollout/on_policy_distillation.py#L312-L329

Training then detaches that scalar, subtracts it from a `[T]` advantage, and uses the ordinary sampled-token policy loss:
https://github.com/radixark/miles/blob/ca25d697c0743fde52d27a5ebd88a3883b7fd838/miles/backends/training_utils/loss_hub/opd.py#L33-L58

At a fixed state, let `c(s)` be that detached scalar. Its immediate on-policy contribution is

```text
E_{a ~ pi_theta(.|s)}[c(s) grad log pi_theta(a|s)]
= c(s) grad sum_a pi_theta(a|s)
= 0.
```

By contrast, differentiating the subset KL gives nonzero gradients for all selected logits whenever the student and teacher distributions differ.

### 4. Minimal regression criterion

A focused test can use three selected logits, compute the renormalized subset reverse KL, call `backward()`, and assert that all three student logits receive gradients and that the teacher-preferred token's gradient increases its probability. The current detached-scalar/sampled-action path cannot satisfy that criterion.

Therefore, the issue is not that using an advantage API is inherently wrong. The official implementation also uses one. The issue is specifically that Miles reduces away the K dimension before training, so the selected-token distribution never enters the train-time autograd graph.
